### ISSUE TITLE
Make block types an own form type (Reusable blocks)

### ISSUE THREAD

**Issue by alexander-schranz** [sulu/sulu#6381]
# Problem description

When managing content it sometimes is the case that a content manager would like to reuse a block with the same content on another page, without having to duplicate it and so to change it on every page again. Currently the only way to achieve this is to tell the developer to create a new block type which uses reusable snippets. With this new global block reusable developer not longer need to ask if something need to be a snippet or not as the control about reusable things are given to content manager.

Another issue with the current implementation is that every block types is injected into every template currently using `xi:includes` this make our metadata a lot bigger as when we make the block types something more central so the frontend only need to load a block types metadata once.

# Proposed solution

Currently blocks types are not a global thing every template can define there own blocks. But from content manager perspective it would be great that a block type is something global.

So blocks can example be marked as `reusable` block and so can be embedded into another page and any change in the reusable block will be update the block in central storage. This would be in my opinion replace the current snippet feature.  

This new blocks could look like the following:

```xml
<!-- config/templates/pages/default.xml

<!-- ... -->
<block name="blocks" default-type="text-image" minOccurs="0">
    <meta>
        <title lang="de">Inhalte</title>
        <title lang="en">Content</title>
    </meta>

    <params>
        <param name="types" type="collection">
            <param name="text-image"/>
            <param name="gallery"/>
            <param name="quote"/>
        </param>
    </params>
</block>

<!-- or property syntax -->
<property name="blocks" type="block">
    <meta>
        <title lang="de">Inhalte</title>
        <title lang="en">Content</title>
    </meta>

    <params>
        <param name="default-type" value="text-image">
        <param name="minOccurs" value="0">
        <param name="types" type="collection">
            <param name="text-image"/>
            <param name="gallery"/>
            <param name="quote"/>
        </param>
    </params>
</block>

<!-- ... -->
```

```xml
<!-- config/templates/blocks/text-image.xml -->
<?xml version="1.0" ?>
<form xmlns="http://schemas.sulu.io/template/template"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/form-1.0.xsd"
>
    <key>text-image</key>

    <properties>
         <!-- ... -->
    </properties>
</form>
```

### Content vs Entity Blocks

Where from content perspective it always make sense that the root block types are a central blocks. There are still valid use cases where the `types` syntax is still valid. The first one is when a block is used inside an entity or a custom model. In this case the data is not saved by sulu and so this can not be reusable blocks or even you don't want to split there the block types into seperate files. So the `<types>` syntax should still be a valid option for data management. Another valid option where the `<types>` syntax still make sense is as example when used inside `nested` blocks as an example we have a block type called `services` which is a block which has against block into it. In this case a `service-item` block type would be an overhead from DX perspective and would make the services.xml block type not really understandable. So a something like this is still valid:

```xml
<!-- config/templates/blocks/text-image.xml -->
<?xml version="1.0" ?>
<form xmlns="http://schemas.sulu.io/template/template"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://schemas.sulu.io/template/template http://schemas.sulu.io/template/form-1.0.xsd"
>
    <key>services</key>

    <properties>
         <property name="article" type="text_line">
              <meta>
                  <title lang="de">Artikel</title>
                  <title lang="en">Article</title>
              </meta>
         </property>

        <block name="services" default-type="service-item" minOccurs="0">
            <meta>
                <title lang="de">Inhalte</title>
                <title lang="en">Content</title>
            </meta>
        
            <params>
                <param name="add_button_text">
                    <meta>
                        <title lang="en">Add service</title>
                    </meta>
                </param>
            </params>
        
            <types>
                <type name="service-item">
                    <properties>
                        <property name="title" type="text_line" colspan="9">
                            <meta>
                                <title lang="de">Artikel</title>
                                <title lang="en">Article</title>
                            </meta>
        
                            <tag name="sulu.block_preview" priority="1024"/>
                        </property>
                        
                        <property name="images" type="single_media_selection">
                            <meta>
                                <title lang="de">Bilder</title>
                                <title lang="en">Images</title>
                            </meta>
        
                            <tag name="sulu.block_preview" priority="512"/>
                        </property>
                    </properties>
                </type>
            </types>
        </block>
    </properties>
</form>
```

**Comment by mavuio** [comment]
✔  i understand the part which talks about having a central place where to define block types without using xi:includes  and i appreciate that 

what i didn't really find in your text  is how the  actual reuse of reusable blocks could work:

where is a block-instance to be marked as reusable ? 

A) does it live in any page and the editor just ticks  a "make-reusable" checkbox 
and then it can be selected in a dropdown at target-destination like a snippet can now ?

how would that element be named or listed in that dropdown ?

B) my background:

i've built a similar feature in another cms based on copying blocks and pasting them as aliases
in arbitrary locations (much like symlinks work in a filesystem)
![clipping-2021_12_22-11_53_45](https://user-images.githubusercontent.com/1092088/147081763-2d1c02c6-edc6-43f8-babe-3dd9f551adc3.png)

![clipping-2021_12_22-11_52_08](https://user-images.githubusercontent.com/1092088/147081563-65611c46-5903-4039-9ed2-22f95f28d67e.png)

the editor has to go to edit the source-instance to change the block's content ( i found it to dangerous to allow it to be edited in all target-destinations and have an effect on all the other instances)

i do not mark the source-block as reusable though, which causes empty references when deleting the source page or -block (much like a symlink would do, but i found editors knew that concept from their windows-shortcut-icons) 

block type-definitions are global there as well, i just define a list of allowed blocks per block-list-definition.

i'm working on a copy-paste feature for sulu anyway, so copying as reference could be integrated in that UX:
https://github.com/sulu/sulu/issues/6436


➜ typo3 has a similar content-reference feature, it features both ways of creating such a reference:
selecting or copying

https://www.typo3manual.com/typo3-manual-6/6-create-reference/61-creating-a-reference-element/

**Comment by spackmat** [comment]
Sulu 2.5 introduced a very TYPO3-ish way to copy/paste blocks that me and my content managers very much appreciate. Thank you, we missed that before! But one aspect is still missing here: Referencing blocks from other pages instead of copy/move them. As described in part here. That is a great feature of TYPO3 as @mavuio already pointed out. Me and my content managers are used to the thought "well, I wrote that already on that other page and want to use the same content here again and without maintaining it at multiple places".

So from my perspective it would be a cool solution if content managers could go to that other page, find the wanted block, click on "copy", go to the new page and instead of inserting a copy just click on reference here from clipboard. That is how I am used to in TYPO3 and that would match the UX of the great new cut/copy/paste-feature. Sulu is almost there.

Or, doing it the other way around: Have a smart content block with a block finder that can be used to select another page and a specific block on it. My first thought on this was to implement that (before Sulu 2.5), but I didn't find a simple solution to list and select certain blocks, only whole pages. In the background that smart content would just use the UUID of the block and load it from the CR, so the minimal approach would be to show the UUID of a block, copy that and feed a very simple smart content block that UUID, so it can load the correct block. Maybe the implementation could just create and provision such a smart content, when a content manager clicks on "paste as reference".

In both variants, the UI must ensure that the reference can only be inserted, when the current page template allows that block type at that place, just as the cut/copy/paste feature (and TYPO3) does. (Or should do, I didn't try it out yet 😉)

**Comment by alexander-schranz** [comment]
@spackmat Thx for your feedback and thx for your interest in this issue.

First I want to tell you the current state. To share content between 2 pages Sulu introduces in its early versions "Snippets" and a content manager can create snippets and a content manager can embed them via "snippet_selection" and "single_snippet_selection".

This issue is about what you did mention in your comment and should target the problem you did list. But at current state there need to be done first an important thing which is the main part of thie issue.

The current block types are defined currently in the templates itself. Which make it at current case possible that a block type "text_image" could be different in template A and different in template B. By making the block types itself a "form type", we are forcing that every "text_image" type is the same. The referencing inside page I currently have the following concept in mind, because we do't want that every block is available verywhere:

```xml
<block name="blocks" default-type="text_image" minOccurs="0">
    <!-- ... -->

    <params>
        <param name="types" type="collection">
            <param name="text_image" />
            <param name="gallery" />
            <param name="embed" />
            <param name="code" />
        </param>
    </params>
</block>
```

This way no xincludes need longer to be used and `text_image` block type is in all pages the same. This is the first important step.

This opens the world for new features. Like the one you did mention which I personally would call "Global Blocks", so blocks which I can mark as "Global" and reference and so I can select a "Global Block" inside another page. And if I change that "Block" it is changed on every page. The UI/UX here we will be needed to be discussed with our partner UI/UX experts. Similar features also exist in current wordpress version.

At the end the question will then more be if this feature would deprecate the existing "snippets" and that "snippets" will be converted into "blocks" or if they can live side by side.

I hope I could give you an overview about current state of content sharing and how this feature should change it.

**Comment by spackmat** [comment]
I don't fully understand the problem, because from my understanding that would affect the new cut/copy/paste feature as well. I mean, the problem you described that it is unclear if a certain block type is the same across different page templates, should also apply to that feature and is somehow solved there. Can't inserting a special reference block for loading a certain block from another page (at least of the same template) reuse that logic? Of course it would be better to make the changes you described, but for now, IMHO it would be enough to just mirror an block of an allowed type over instead of pasting a copy of it.

Edit: I think what you call "global blocks" would be very close to snippets. But simply having a "reference" block type like in TYPO3 would be a different thing, so that and snippets can very well coexist.

**Comment by alexander-schranz** [comment]
> I don't fully understand the problem, because from my understanding that would affect the new cut/copy/paste feature as well.

The cut/copy/paste function is not effected. If it can map the data it does map the data if not it will use the "default type" and if the default type has some fields in it it can map them and all others are ignored and the content manager paste things and that pasted things the content manager will click save and if the data is not valid he need to change or add something if there are not all required fields now field for paste data.

> IMHO it would be enough to just mirror an block of an allowed type over instead of pasting a copy of it.

That would be just a fast shoot into a wrong direction from my point of view. But this issue was not yet in deep discussed with the team and is only my point of view on this topic. But I don't think we will just implement a reference type, as that would make other problems like side effects of removed pages changing other pages and that is why I would direct reference avoid and go over converting a block into a global block.

**Comment by spackmat** [comment]
Now I see, thanks! Interesting approach for the implementation of the paste feature BTW.

**Comment by stollr** [comment]
As a new Sulu user I can confirm that this is a feature I searched for, when I started using Sulu some weeks ago. I was a little bit disappointed when I found out that this wasn't supported. Later I detected the xinclude feature, which is some kind of workaround. But I'd be really happy if this feature will be implemented one day ;-)

**Comment by alexander-schranz** [comment]
@naitsirch Yes sharing the form structure between 2 templates the `xi:include` feature is still the way to go. It can be used for any kind of xml data not only blocks also for sections.

This feature is then more about not sharing block form metadata instead sharing also whole content. And make it possible for the content manager change a reuseable block which can be used on different pages. But step by step first step will be that a block itself need to be a more central form metadata :)

-------------

@luca-rath What do you think we need to keep in mind in case of the JSON validation implementation, as the block content type would need to include form metadata of the block somewhere to great the overall json validaton.

**Comment by luca-rath** [comment]
@alexander-schranz I can think of two different examples of how reusable blocks could be nicely implemented, but I don't think that would affect the JSON Schema implementation a lot.

## Option 1

```xml
<!-- config/blocks/text-with-image.xml -->
<type name="text-with-image">
    <properties>
        <!-- ... -->
    </properties>
</type>
```

```xml
<!-- config/pages/page.xml -->
<template>
    <key>page</key>
    <!-- ... -->
    <properties>
        <block>
            <types>
                <type ref="text-with-image"/>
                <type name="some-local-type">
                    <properties>
                        <!-- ... -->
                    </properties>
                </type>
            </types>
        </block>
    </properties>
</type>
```

(I would personally prefer an approach like this this one)


## Option 2

```xml
<!-- config/pages/page.xml -->
<template>
    <key>page</key>
    <!-- ... -->
    <properties>
        <block>
            <types>
                <!-- In this example, just one template is allowed to define a block with the same name and the `global` flag -->
                <type name="text-with-image" global="true">
                    <properties>
                        <!-- ... -->
                    </properties>
                </type>

                <type name="some-local-type">
                    <properties>
                        <!-- ... -->
                    </properties>
                </type>
            </types>
        </block>
    </properties>
</type>
```

```xml
<!-- config/pages/page-2.xml -->
<template>
    <key>page-2</key>
    <!-- ... -->
    <properties>
        <block>
            <types>
                <type ref="text-with-image"/>
            </types>
        </block>
    </properties>
</type>
```

I think in both cases, it doesn't matter, where the json schema comes from. A `PropertyMetadataMapper` for the `Block` field type needs to be responsible to load the schema of it's defined/referenced types and compose those schemata together in an `AllOfsMetadata`, like it's currently done in the `FormMetadataMapper`

**Comment by alexander-schranz** [comment]
@luca-rath it will be global like in having an `config/templates/content-blocks/*.xml` as they need a `global` unique key. In the `block` content type they will be maybe referenced this way:

```xml
<property name="blocks" type="block">
    <params>
        <param name="types" type="collection">
            <param name="text-image"/>
            <param name="gallery"/>
            <param name="quote"/>
        </param>
    </params>
</property>
```

> I think in both cases, it doesn't matter, where the json schema comes from. A PropertyMetadataMapper for the Block field type needs to be responsible to load the schema of it's defined/referenced types and compose those schemata together in an AllOfsMetadata, like it's currently done in the FormMetadataMapper

Okay splitting and including something like `$ref` you think should work. Like in this example: https://stackoverflow.com/questions/18376215/jsonschema-split-one-big-schema-file-into-multiple-logical-smaller-files

So we maybe only need to load the schemas which are required for us. And avoid one big schema which was also issue in the past in big projects having mb size of page template metadata data.

**Comment by luca-rath** [comment]
> @luca-rath it will be global like in having an `config/templates/content-blocks/*.xml` as they need a `global` unique key. In the `block` content type they will be maybe referenced this way:

Yeah I saw after answering ... I didn't read the issue carefully before 😂

at least ajv supports that, yes

**Comment by luca-rath** [comment]
> at least ajv supports that, yes

Our own json schema implementation on the server side doesn't support it yet, but that shouldn't be a big deal

**Comment by mamazu** [comment]
@chirimoya This should be done with your PR and this can be closed?
